User parameter handling with float vectors

Turns out to be tricky. See comment from support.hpp

```
// This is not actually really enough to work generally as there's an
// issue with what to do with checking previous, min and max against
// NA_REAL -- which is not going to be the correct value for float
// rather than double.  Further, this is not extendable to the vector
// cases because we hit issues around partial template specification.
//
// We can make the latter go away by replacing std::array<T, N> with
// std::vector<T> - the cost is not great.  But the NA issues remain
// and will require further thought. However, this template
// specialisation and the tests that use it ensure that the core code
// generation is at least compatible with floats.
template <>
float user_get_scalar<float>(Rcpp::List user, const char *name,
                             const float previous, float min, float max);
```

Trigger an error with

```r
gen <- odin_dust_("examples/array.R", real_t = "float", verbose = TRUE)
```

```
ℹ 14 functions decorated with [[cpp11::register]]
✔ generated file cpp11.R
✔ generated file cpp11.cpp
ℹ 14 functions decorated with [[cpp11::register]]
✔ generated file cpp11.R
✔ generated file cpp11.cpp
Re-compiling arrayde799afc
─  installing *source* package ‘arrayde799afc’ ...
   ** using staged installation
   ** libs
   g++ -std=gnu++11 -I"/usr/share/R/include" -DNDEBUG  -I'/home/rfitzjoh/lib/R/library/cpp11/include' -g -Wall -Wextra -pedantic -Wmaybe-uninitialized -Wno-unused-parameter -Wno-unused-variable -Wno-cast-function-type -Wno-missing-field-initializers -O2  -I/home/rfitzjoh/lib/R/library/dust/include -DHAVE_INLINE -fopenmp -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-8T8CYO/r-base-4.0.3=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c cpp11.cpp -o cpp11.o
   g++ -std=gnu++11 -I"/usr/share/R/include" -DNDEBUG  -I'/home/rfitzjoh/lib/R/library/cpp11/include' -g -Wall -Wextra -pedantic -Wmaybe-uninitialized -Wno-unused-parameter -Wno-unused-variable -Wno-cast-function-type -Wno-missing-field-initializers -O2  -I/home/rfitzjoh/lib/R/library/dust/include -DHAVE_INLINE -fopenmp -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-8T8CYO/r-base-4.0.3=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c dust.cpp -o dust.o
   dust.cpp: In function ‘cpp11::sexp dust_info(const typename T::init_t&) [with T = array; typename T::init_t = array::init_t]’:
   dust.cpp:316:27: warning: variable ‘internal’ set but not used [-Wunused-but-set-variable]
     316 |   const array::internal_t internal = init.internal;
         |                           ^~~~~~~~
   dust.cpp: At global scope:
   dust.cpp:69:10: warning: inline function ‘T na_value() [with T = float]’ used but never defined
      69 | inline T na_value();
         |          ^~~~~~~~
   g++ -std=gnu++11 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o arrayde799afc.so cpp11.o dust.o -fopenmp -L/usr/lib/R/lib -lR
   installing to /tmp/Rtmp0zju0f/devtools_install_11b4efcf681d/00LOCK-file11b4e6b2b9eec/00new/arrayde799afc/libs
   ** checking absolute paths in shared objects and dynamic libraries
─  DONE (arrayde799afc)
Error in dyn.load(dll) : 
  unable to load shared object '/tmp/Rtmp0zju0f/file11b4e6b2b9eec/src/arrayde799afc.so':
  /tmp/Rtmp0zju0f/file11b4e6b2b9eec/src/arrayde799afc.so: undefined symbol: _ZN5cpp118r_vectorIfE10valid_typeEP7SEXPREC
```

--- a/DESCRIPTION
+++ b/DESCRIPTION
@@ -1,6 +1,6 @@
 Package: odin.dust
 Title: Compile Odin to Dust
-Version: 0.1.5
+Version: 0.1.6
 Authors@R: c(person("Rich", "FitzJohn", role = c("aut", "cre"),
                     email = "rich.fitzjohn@gmail.com"),
              person("John", "Lees", role = "aut"),

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,6 +1,6 @@
 # odin.dust 0.1.6
 
-* Support for vectors of floats (rater than doubles) (#6)
+* Support for vectors of floats (rather than doubles) (#6)
 
 # odin.dust 0.1.5
 

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,3 +1,7 @@
+# odin.dust 0.1.6
+
+* Support for vectors of floats (rater than doubles) (#6)
+
 # odin.dust 0.1.5
 
 * Support `as.integer()` in odin code (since odin 1.0.7) (#43)

--- a/inst/support.hpp
+++ b/inst/support.hpp
@@ -7,21 +7,6 @@
 #include <cpp11/strings.hpp>
 #include <vector>
 
-// These would be nice to make constexpr but the way that NA values
-// are defined in R's include files do not allow it.
-template <typename T>
-inline T na_value();
-
-template <>
-inline int na_value<int>() {
-  return NA_INTEGER;
-}
-
-template <>
-inline double na_value<double>() {
-  return NA_REAL;
-}
-
 template <typename T>
 inline bool is_na(T x);
 
@@ -41,13 +26,13 @@ inline size_t object_length(cpp11::sexp x) {
 
 template <typename T>
 void user_check_value(T value, const char *name, T min, T max) {
-  if (ISNA(value)) {
+  if (is_na(value)) {
     cpp11::stop("'%s' must not be NA", name);
   }
-  if (min != na_value<T>() && value < min) {
+  if (!is_na(min) && value < min) {
     cpp11::stop("Expected '%s' to be at least %g", name, (double) min);
   }
-  if (max != na_value<T>() && value > max) {
+  if (!is_na(max) && value > max) {
     cpp11::stop("Expected '%s' to be at most %g", name, (double) max);
   }
 }
@@ -142,24 +127,19 @@ T user_get_scalar(cpp11::list user, const char *name,
   return ret;
 }
 
-// This is not actually really enough to work generally as there's an
-// issue with what to do with checking previous, min and max against
-// NA_REAL -- which is not going to be the correct value for float
-// rather than double.  Further, this is not extendable to the vector
-// cases because we hit issues around partial template specification.
-//
-// We can make the latter go away by replacing std::array<T, N> with
-// std::vector<T> - the cost is not great.  But the NA issues remain
-// and will require further thought. However, this template
-// specialisation and the tests that use it ensure that the core code
-// generation is at least compatible with floats.
-//
-// See #6
 template <>
 inline float user_get_scalar<float>(cpp11::list user, const char *name,
                                     const float previous, float min, float max) {
   double value = user_get_scalar<double>(user, name, previous, min, max);
   return static_cast<float>(value);
+}
+
+template <typename T>
+std::vector<T> user_get_array_value(cpp11::sexp x, const char * name,
+                                    T min, T max) {
+  std::vector<T> ret = cpp11::as_cpp<std::vector<T>>(x);
+  user_check_array_value<T>(ret, name, min, max);
+  return ret;
 }
 
 template <typename T, size_t N>
@@ -178,10 +158,7 @@ std::vector<T> user_get_array_fixed(cpp11::list user, const char *name,
   user_check_array_rank<N>(x, name);
   user_check_array_dim<N>(x, name, dim);
 
-  std::vector<T> ret = cpp11::as_cpp<std::vector<T>>(x);
-  user_check_array_value(ret, name, min, max);
-
-  return ret;
+  return user_get_array_value<T>(x, name, min, max);
 }
 
 template <typename T, size_t N>
@@ -200,9 +177,17 @@ std::vector<T> user_get_array_variable(cpp11::list user, const char *name,
   user_check_array_rank<N>(x, name);
   user_set_array_dim<N>(x, name, dim);
 
-  std::vector<T> ret = cpp11::as_cpp<std::vector<T>>(x);
-  user_check_array_value(ret, name, min, max);
+  return user_get_array_value<T>(x, name, min, max);
+}
 
+template <>
+inline std::vector<float> user_get_array_value(cpp11::sexp x, const char * name,
+                                               float min, float max) {
+  // NOTE: possible under/overflow here for min/max because we've
+  // downcast this.
+  std::vector<double> value = user_get_array_value<double>(x, name, min, max);
+  std::vector<float> ret(value.size());
+  std::copy(value.begin(), value.end(), ret.begin());
   return ret;
 }
 

--- a/tests/testthat/test-odin-dust.R
+++ b/tests/testthat/test-odin-dust.R
@@ -368,6 +368,28 @@ test_that("sir model float test", {
 })
 
 
+test_that("array model float test", {
+  gen_f <- odin_dust_("examples/array.R", real_t = "float", verbose = FALSE)
+  gen_d <- odin_dust_("examples/array.R", real_t = "double", verbose = FALSE)
+
+  r <- matrix(runif(10), 2, 5)
+  x0 <- matrix(runif(10), 2, 5)
+
+  mod_f <- gen_f$new(list(x0 = x0, r = r), 0, 1)
+  mod_d <- gen_d$new(list(x0 = x0, r = r), 0, 1)
+
+  expect_identical(mod_d$state(), matrix(c(x0)))
+  expect_equal(mod_f$state(), mod_d$state(), tolerance = 1e-7)
+  expect_false(identical(mod_f$state(), mod_d$state()))
+
+  y_d <- mod_d$run(1)
+  y_f <- mod_f$run(1)
+  expect_identical(y_d, matrix(c(x0 + r)))
+  expect_equal(y_f, y_d, tolerance = 1e-7)
+  expect_false(identical(y_f, y_d))
+})
+
+
 test_that("specify workdir", {
   path <- tempfile()
   gen <- odin_dust({

